si-timers/src/Control/Monad/Class/MonadTimer/SI.hs

```diff
@@ -60,15 +65,24 @@ class ( MonadTimer.MonadDelay m
       ) => MonadDelay m where
   threadDelay :: DiffTime -> m ()
 
--- | Thread delay.  When the delay is smaller than what `Int` can represent it
--- will use the `Control.Monad.Class.MonadTimer.threadDelay` (e.g. for the `IO`
--- monad it will use `Control.Concurrent.threadDelay`); otherwise it will
--- recursively call `Control.Monad.Class.MonadTimer.threadDelay`.
+-- | Thread delay. This implementation will not over- or underflow.
+--
+-- For delay larger than what `Int` can represent (see
+-- `diffTimeToMicrosecondsAsInt`), it will recursively call
+-- `Control.Monad.Class.MonadTimer.threadDelay`.
+--
+-- For delays smaller than `minBound :: Int` seconds, `minBound :: Int` will be
+-- used instead.
 --
 instance MonadDelay IO where
   threadDelay :: forall m.
                  MonadDelay m
               => DiffTime -> m ()
+  threadDelay d | d <= minDelay =
+      MonadTimer.threadDelay minBound
+    where
+      minDelay :: DiffTime
+      minDelay = microsecondsAsIntToDiffTime minBound
```

Comment:
1) `io-classes:threadDelay` should have the same behavior as `IO.threadDelay`. I _think_ that's a `yield` for negative durations, but I don't know. Just calling `IO.threadDelay` with a negative number avoids any mismatch.

2) My use case that involved negative durations was some code that needed to delay "unless enough time had already passed", and `threadDelay (dur - sofar)` was the easiest way to do that. I had some extreme `sofar` numbers in my tests, and saw an effectively-interminable delay by surprise.

Reply:
I am not sure how `threadDelay 0` behaves, but either this or or:
```hs
threadDelay d | d < 0 = return ()
```
is fine for me.  If people change `threadDelay` behaviour a lot of code will be left broken, so I don't think it will ever change.

Reply:
@njd42 is right, it's non-positive timeouts that are immediately discharged, see [registerTimeout](https://hackage.haskell.org/package/base-4.19.0.0/docs/src/GHC.Event.TimerManager.html#registerTimeout).

Our `io-sim` implementation of `threadDelay 0` has a different semantics: it will put the thread on back of the run queue.